core/src/main/scala/com/nvidia/spark/rapids/tool/profiling/Analysis.scala

```diff
@@ -37,180 +38,127 @@ class Analysis(apps: Seq[ApplicationInfo]) {
     }
   }
 
-  private def maxWithEmptyHandling(arr: ArrayBuffer[Long]): Long = {
+  private def maxWithEmptyHandling(arr: Iterable[Long]): Long = {
     if (arr.isEmpty) {
       0L
     } else {
       arr.max
     }
   }
 
+  private def minWithEmptyHandling(arr: Iterable[Long]): Long = {
+    if (arr.isEmpty) {
+      0L
+    } else {
+      arr.min
+    }
+  }
+
   // Job + Stage Level TaskMetrics Aggregation
   def jobAndStageMetricsAggregation(): Seq[JobStageAggTaskMetricsProfileResult] = {
-    val allJobRows = apps.flatMap { app =>
-      app.jobIdToInfo.map { case (id, jc) =>
-        val stagesInJob = app.stageIdToInfo.filterKeys { case (sid, _) =>
-          jc.stageIds.contains(sid)
-        }.keys.map(_._1).toSet
-        if (stagesInJob.isEmpty) {
-          None
-        } else {
-          val tasksInJob = app.taskEnd.filter { tc =>
-            stagesInJob.contains(tc.stageId)
-          }
-          // count duplicate task attempts
-          val (durSum, durMax, durMin, durAvg) = getDurations(tasksInJob)
-          Some(JobStageAggTaskMetricsProfileResult(app.index,
-            s"job_$id",
-            tasksInJob.size,
-            jc.duration,
-            tasksInJob.map(_.diskBytesSpilled).sum,
-            durSum,
-            durMax,
-            durMin,
-            durAvg,
-            tasksInJob.map(_.executorCPUTime).sum,
-            tasksInJob.map(_.executorDeserializeCPUTime).sum,
-            tasksInJob.map(_.executorDeserializeTime).sum,
-            tasksInJob.map(_.executorRunTime).sum,
-            tasksInJob.map(_.input_bytesRead).sum,
-            tasksInJob.map(_.input_recordsRead).sum,
-            tasksInJob.map(_.jvmGCTime).sum,
-            tasksInJob.map(_.memoryBytesSpilled).sum,
-            tasksInJob.map(_.output_bytesWritten).sum,
-            tasksInJob.map(_.output_recordsWritten).sum,
-            maxWithEmptyHandling(tasksInJob.map(_.peakExecutionMemory)),
-            tasksInJob.map(_.resultSerializationTime).sum,
-            maxWithEmptyHandling(tasksInJob.map(_.resultSize)),
-            tasksInJob.map(_.sr_fetchWaitTime).sum,
-            tasksInJob.map(_.sr_localBlocksFetched).sum,
-            tasksInJob.map(_.sr_localBytesRead).sum,
-            tasksInJob.map(_.sr_remoteBlocksFetched).sum,
-            tasksInJob.map(_.sr_remoteBytesRead).sum,
-            tasksInJob.map(_.sr_remoteBytesReadToDisk).sum,
-            tasksInJob.map(_.sr_totalBytesRead).sum,
-            tasksInJob.map(_.sw_bytesWritten).sum,
-            tasksInJob.map(_.sw_recordsWritten).sum,
-            tasksInJob.map(_.sw_writeTime).sum
-          ))
+    // first get all stage aggregated levels
```

Comment:
Old code used to iterate on all jobs to get the stages, then iterate on all tasks within each stage to aggregate. This will create all the jobs rows.
Then it will do the same sequence to get all the stage rows.
This is clearly very time consuming.
Instead the new code does the following:

- Loop on all the stages and aggreate them.
- Cache the results in a hashMap.
- Loop on all the jobs and aggregate the values cached within the hashMap.